src/StaticLS/IDE/CodeActions/AutoImport.hs

```diff
@@ -43,28 +45,65 @@ type AutoImportTypes =
     :+ Haskell.ConstructorOperator
     :+ Nil
 
+data ModulesToImport =
+  ModulesToImport
+    {
+      moduleNames :: [Text]
+    , moduleQualifier :: Maybe Text
+    }
+
 getModulesToImport ::
   (HasCallStack, ()) =>
   AbsPath ->
   LineCol ->
-  StaticLsM [Text]
+  StaticLsM ModulesToImport
 getModulesToImport path pos = do
   _ <- logInfo "getModulesToImport"
   haskell <- getHaskell path
 
   let astPoint = lineColToAstPoint pos
   _ <- logInfo $ T.pack $ "astPoint: " ++ show astPoint
   _ <- logInfo "got haskell"
-  let maybeQualified = AST.getDeepestContaining @AutoImportTypes astPoint (AST.getDynNode haskell)
-  case maybeQualified of
-    Just qualified -> do
-      let node = AST.getDynNode qualified
-      let nodeText = AST.nodeText node
-      _ <- logInfo $ T.pack $ "qualified: " ++ show node
-      _ <- logInfo $ T.pack $ "qualified: " ++ show nodeText
+  let maybeQualified =
+        AST.getDeepestContaining
```

Comment:
This does result in double traversing the AST unfortunately - maybe could use a "getAllMatchingContaining" upstream to avoid this?